### ISSUE TITLE
CheckFinite() failure (in heat method distance algorithm) for some .off datasets

### ISSUE THREAD

**Issue by ItakEjgo** [nmwsharp/geometry-central#36]
Dear nmwsharp, 
I am currently trying to use the **heat method distance algorithm** in GC for some .off datasets. Here is the link of datasets I used:
[https://github.com/ItakEjgo/TerrainDataset](url)
However, I met some problems. I follow your gc tutorial for the heat method distance algorithm through the following code: 
```
HeatMethodDistanceSolver heatSolver(*geometry);
for (decltype(mesh->nVertices()) i = 0; i < mesh->nVertices(); i++){
    Vertex v = mesh->vertex(i);
    VertexData<double> distToSource = heatSolver.computeDistance(v);
}
```
This algorithm wells really well for some datasets and performs really well, e.g. small_terrain.off and BH_low.off.
But, it failed in the other .off datasets with the following exception:
```
terminate called after throwing an instance of 'std::logic_error'
  what():  checkFinite() failure. Non-finite row vector entry [39] = -nan
Aborted (core dumped)
```
The value of the vector became **-nan** and thus the checkFinite() function throws the previous exception for these datasets. I tried to fix this problem and found that the problem is happened in the following line in **heat_method_distance.cpp line 174**:
`Vector<double> distVec = poissonSolver->solve(divergenceVec);`
But I have no idea about solving this problem...
All these dataset could be successfully visulized through the **mesh lab** in [https://github.com/cnr-isti-vclab/meshlab](url).
Could you kindly give me some adivce about how to solve this problem？Are the datasets I used are vaild or invalid? 
Thanks in advance for your help!

**Comment by nmwsharp** [comment]
Hi, thanks for posting!

This is probably happening because your meshes have some degenerate faces: extremely skinny zero- or nearly-zero area triangles. This leads to bad numerics in the "cotangent" weights that show up in the Laplace and divergence operator for that algorithm, eventually manifesting as `NaN` and `inf` values. 

One solution would be to try to 'repair' these skinny triangles, e.g. using MeshLab or some other software. However, this means messing with your data, and sometimes the repair process can create other new problems. 

Fortunately, geometry-central has facilities for dealing with exactly this problem without remeshing (and this is something I do research on!). One good solution is to use _intrinsic mollification_, as shown here: http://geometry-central.net/surface/algorithms/robust_geometry/. 

You can convert your `VertexPositionGeometry` to an `EdgeLengthGeometry` and use intrinsic mollification with something like:

```cpp
geometry->requireEdgeLengths();
EdgeLengthGeometry intrinsicGeometry(*mesh, geometry->edgeLengths);

// mollify
mollifyIntrinsic(*mesh, intrinsicGeometry.inputEdgeLengths);

// create the heat solver as before
HeatMethodDistanceSolver heatSolver(intrinsicGeometry);

// same as before
```  

You can get even better numerical accuracy by using "intrinsic edge flips" in addition to mollification.... just let me know if you're interested in that!

**Comment by ItakEjgo** [comment]
> Hi, thanks for posting!
> 
> This is probably happening because your meshes have some degenerate faces: extremely skinny zero- or nearly-zero area triangles. This leads to bad numerics in the "cotangent" weights that show up in the Laplace and divergence operator for that algorithm, eventually manifesting as `NaN` and `inf` values.
> 
> One solution would be to try to 'repair' these skinny triangles, e.g. using MeshLab or some other software. However, this means messing with your data, and sometimes the repair process can create other new problems.
> 
> Fortunately, geometry-central has facilities for dealing with exactly this problem without remeshing (and this is something I do research on!). One good solution is to use _intrinsic mollification_, as shown here: http://geometry-central.net/surface/algorithms/robust_geometry/.
> 
> You can convert your `VertexPositionGeometry` to an `EdgeLengthGeometry` and use intrinsic mollification with something like:
> 
> ```c++
> geometry->requireEdgeLengths();
> EdgeLengthGeometry intrinsicGeometry(*mesh, geometry->edgeLengths);
> 
> // mollify
> mollifyIntrinsic(*mesh, intrinsicGeometry.inputEdgeLengths);
> 
> // create the heat solver as before
> HeatMethodDistanceSolver heatSolver(intrinsicGeometry);
> 
> // same as before
> ```
> 
> You can get even better numerical accuracy by using "intrinsic edge flips" in addition to mollification.... just let me know if you're interested in that!

I just follow your advice and it turns out to work well after including an extra header:

`#include "geometrycentral/surface/edge_length_geometry.h"`

 Now the algorithm could run on my datasets. Thanks for your kindly advice! 

What's more, I am really interested about the “intrinsic edge filps”, could you give me some references about that?

**Comment by ItakEjgo** [comment]
> Hi, thanks for posting!
> 
> This is probably happening because your meshes have some degenerate faces: extremely skinny zero- or nearly-zero area triangles. This leads to bad numerics in the "cotangent" weights that show up in the Laplace and divergence operator for that algorithm, eventually manifesting as `NaN` and `inf` values.
> 
> One solution would be to try to 'repair' these skinny triangles, e.g. using MeshLab or some other software. However, this means messing with your data, and sometimes the repair process can create other new problems.
> 
> Fortunately, geometry-central has facilities for dealing with exactly this problem without remeshing (and this is something I do research on!). One good solution is to use _intrinsic mollification_, as shown here: http://geometry-central.net/surface/algorithms/robust_geometry/.
> 
> You can convert your `VertexPositionGeometry` to an `EdgeLengthGeometry` and use intrinsic mollification with something like:
> 
> ```c++
> geometry->requireEdgeLengths();
> EdgeLengthGeometry intrinsicGeometry(*mesh, geometry->edgeLengths);
> 
> // mollify
> mollifyIntrinsic(*mesh, intrinsicGeometry.inputEdgeLengths);
> 
> // create the heat solver as before
> HeatMethodDistanceSolver heatSolver(intrinsicGeometry);
> 
> // same as before
> ```
> 
> You can get even better numerical accuracy by using "intrinsic edge flips" in addition to mollification.... just let me know if you're interested in that!

Dear nmwsharp,
Thanks for your fast reply and kindly help!

I followed your advice with including an extra header:
`#include "geometrycentral/surface/edge_length_geometry.h"`
However, the heat distance method still fails with some points as the source (in SF_low.txt). Here some results from my testing code:
```
mesh load finished.
heat solver process finished.
node 0 failed to compute
node 1 failed to compute
node 2 failed to compute
node 3 time cost is 0 second(s)
node 4 time cost is 0 second(s)
node 5 failed to compute
```

The int value after "node" is the index of vertex. For some vertices, the algorithm works well with fast response. But it still fails for the other vertices. I use the following code to generate EdgeLengthGeometry: 

```
std::unique_ptr<SurfaceMesh> mesh;
std::unique_ptr<VertexPositionGeometry> geometry;

std::tie(mesh, geometry) = loadMesh(argv[1]);
std::cout << "mesh load finished." << std::endl;

geometry->requireEdgeLengths();
EdgeLengthGeometry intrinsicGeometry(*mesh, geometry->edgeLengths);
mollifyIntrinsic(*mesh, intrinsicGeometry.inputEdgeLengths);

HeatMethodDistanceSolver heatSolver(intrinsicGeometry);
```

I also tried to set a larger **relativeFactor** in the mollifyIntrinsic function ( 1e-4 I used), but it still did not work.
Am I using the convertion in the correct way? I wonder whether I made some mistakes in using the mollifyIntrinsic function.
What's more, I am quite interested in the **"intrinsic edge flips"**, could you give me some references about it? 

Thanks a lot!

**Comment by nmwsharp** [comment]
Hi!

OK, I downloaded your data and looked at this a bit more.

In your case, the culprit was not bad numerics from skinny triangles, but rather numerical underflow. The heat method for distance algorithm uses short time heat flow to compute distance, and in your case the meshes are large enough that the solution to heat flow was getting numerical underflow, even in double precision. This underflow was leading all `0` heat values in far-away regions of the mesh. This caused a `0`-vector to get normalized and produce `NaN`s. It just happened to work out for some vertices which were near the middle of the mesh, since no point on the domain was too far away.

I just pushed a small fix to the master branch (3b31f8b848bb2c440c98a71a1e699372a251c20c) which adds a threshold to the normalization so that the code doesn't crash and burn with `NaN`s  in this case. However, the computed distance values still are not correct in distant regions. 

To actually fix the problem, you can change the time coefficient to some value larger than `1`, like
```cpp
 HeatMethodDistanceSolver heatSolver(intrinsicGeometry, 10.);
```
and it should work well on your meshes. If you see this issue on bigger and bigger meshes, you can keep using larger values for that last time parameter. We don't always use bigger values here because they slightly "smooth out" the resulting solutions. But in your case using somewhat bigger values should be just fine.

Finally, for more references on intrinsic geometry, check out two of our recent papers:

- http://www.cs.cmu.edu/~kmcrane/Projects/NavigatingIntrinsicTriangulations/paper.pdf
- http://www.cs.cmu.edu/~kmcrane/Projects/NonmanifoldLaplace/NonmanifoldLaplace.pdf

The first is a new data structure for representing intrinsic triangulations, but also contains a pretty good general introduction to the idea. The second shows how to extend the technique to general nonmanifold triangulations, and also shows more examples of using intrinsic triangulations to make algorithms more robust.

Best,
Nick

**Comment by ItakEjgo** [comment]
Dear Nick,

Thanks for your valuable reply! I will follow your advice and learn some knowledge about the DDG in: brickisland.net/DDGSpring2020/.

I am really fascinated by this beautiful method and want to do something useful with it. 

Really thanks for your comments and effort! I am so impressed by your work!

Best wishes,
ItakEjgo
